README.md
# Crypto Tools

Crypto tools is a powerful tool which contains lots of encoding/decoding and encryption/decryption tools like base64、hash、urlencode、AES、RSA、Vigenere、Caesar、etc.,and additional provides some text manipulation functions

## Features

> Coding/Decoding

- base16/32/64
- MD5/SHA512
- Url Encode
- Html Entitles
- ROT13
- Quote-Printable
- Bubble Babble
- Brainfuck Interpreter
- Number/Hex/Bin and String Conversion

![](./images/DecodeEncode.gif)

> Encryption/Decryption

- Symmetric Encryption (AES/DES/RC4/etc.)
- RSA Encryption
- Caesar Cipher
- Vigenere Cipher and Guess Key 
- Fence Cipher
- Morse Code
- Cracking Hashes online

![](./images/DecryptEncrypt.gif)

> Text Manipulation

- Reverse String
- All Upper/Lower Case
- Title Case
- Strip
- Space to Line/None
- Convert / and \
- Get String Length
- Add Quote For Each Word

![](./images/TextManipulation.gif)

> Hexadecimal Calculator

- Simple Hexadecimal Calculator

![](./images/HexadecimalConverter.gif)

>Solve the Conflicts with Output Colorizer extension

- Combine the syntaxes files for each installed output colorizer extension

## Requirements

Crack hashes online based on HashBuster,so you need to install python and related libraries 

python2:

```
pip install requests futures
```

python3:

```
pip install requests
```

## Usage

Active the extension as follows:

- Ctrl + shift + P, then just type `crypto` to find the command
- Select a text, then right click to find the command
- Use keybindings when a text has been selected



Default keybindings:

- `Encode/Decode:`          ctrl+alt+d when editorHasSelection
- `Encrypt/Decrypt:`        ctrl+alt+a when editorHasSelection
- `Text Manipulation:`      ctrl+alt+t when editorHasSelection
- `Hexadecimal Calculator:` ctrl+alt+c

You can change it by yourself

## Known Issues

Submit the [issues](https://github.com/fofolee/vscode-crypto-tools/issues) if you find any bug or have any suggestion.

## Release Notes

### 0.0.2

- Solve the Conflicts with Output Colorizer extension

- bug fixes

### 0.0.1

Initial release


CHANGELOG.md
### 0.0.3 (2023.1.4)

 - Fix the base64 encoding bug of Chinese
 - Fix symmetric cipher bug
 - add sm4-ecb
 - add url-safe base64
 - Adjust the position of some algorithms

### 0.0.2 (2019.2.7)

- Solve the Conflicts with Output Colorizer extension
- bug fixes

### 0.0.1 (2019.2.6)

Initial release


scripts/Basic.js
const base32 = require('./base32')
const rot = require('./Caesar').rot
const crypto = require('crypto')
const urlencode = require('urlencode')
const htmlencode = require('htmlencode')
const quopri = require('quoted-printable')
const utf8 = require('utf8')
const bubble = require('bubble_babble')
const brainfuck = require('./brainfuck')

exports.hex2string = function (str) {
  return new Buffer(str, 'hex').toString()
}

exports.string2hex = function (str) {
  return new Buffer(str).toString('hex')
}

exports.number2string = function (str) {
  return exports.hex2string(parseInt(str).toString(16))
}

exports.string2number = function (str) {
  return parseInt(exports.string2hex(str), 16).toString()
}

exports.bin2string = function (str) {
  if (!str.length % 8) return
  var s = ''
  for (var i = 0; i < str.length; i += 8) {
    s += String.fromCharCode(parseInt(str.slice(i, i + 8), 2))
  }
  return s
}

exports.string2bin = function (str) {
  return str
    .split('')
    .map((i) => {
      return i.charCodeAt(0).toString('2').padStart(8, '0')
    })
    .join('')
}

exports.base64Encode = function (str) {
  return new Buffer(str, 'utf8').toString('base64')
}

let base64Decode = function (str) {
  return new Buffer(str, 'base64').toString()
}

exports.urlSafeBase64Encode = function (str) {
  return exports
    .base64Encode(str)
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/\=/g, '')
}

let urlSafeBase64Decode = function (base64Str) {
  base64Str = base64Str.replace(/\-/g, '+').replace(/\_/g, '/')
  let num = base64Str.length % 4
  return base64Str + '===='.substring(0, num)
}

exports.base32Encode = function (str) {
  return base32.encode(str)
}

let base32Decode = function (str) {
  return base32.decode(str)
}

exports.base16Encode = function (str) {
  return exports.string2hex(str).toUpperCase()
}

let base16Decode = function (str) {
  if (/[^0-9A-F]/.test(str)) {
    throw new Error('Invalid base16 characters')
  }
  return exports.hex2string(str)
}

exports.baseDecode = function (str) {
  try {
    return base16Decode(str)
  } catch (error) {
    str = urlSafeBase64Decode(str)
    try {
      return base32Decode(str)
    } catch (error) {
      return base64Decode(str)
    }
  }
}

exports.rot13 = function (str) {
  return rot(str, 13)
}

exports.md5Hash = function (str) {
  let hash = crypto.createHash('md5')
  return hash.update(str).digest('hex')
}

exports.sha512Hash = function (str) {
  let hash = crypto.createHash('sha512')
  return hash.update(str).digest('hex')
}

exports.urlEncode = function (str) {
  return urlencode(str)
}

exports.urlDecode = function (str) {
  return urlencode.decode(str)
}

exports.htmlEncode = function (str) {
  return htmlencode.htmlEncode(str)
}

exports.htmlDecode = function (str) {
  return htmlencode.htmlDecode(str)
}

exports.quopriEncode = function (str) {
  return quopri.encode(utf8.encode(str))
}

exports.quopriDecode = function (str) {
  return utf8.decode(quopri.decode(str))
}

exports.bubbleDecode = function (str) {
  return bubble.decode(str).toString()
}

exports.bubbleEncode = function (str) {
  return bubble.encode(str)
}

exports.brainfuckDecode = function (str) {
  return brainfuck.run(str)
}


scripts/Fence.js
function decrypt(string,rows) {
  let fence = [];
  for (let i = 0; i < rows; i++) fence.push([])
  let rail = 0;
  let change = 1;

  string.split("").forEach(char => {
    fence[rail].push(char)
    rail += change

    if (rail === rows - 1 || rail === 0) change = -change
  })

  const rFence = [];
  for (let i = 0; i < rows; i++) rFence.push([])

  i = 0
  s = string.split("")
  for (r of fence) {
    for (let j = 0; j < r.length; j++) rFence[i].push(s.shift())
    i++
  }

  rail = 0
  change = 1
  var r = ""
  for (var i = 0; i < string.length; i++) {
    r += rFence[rail].shift()
    rail += change

    if (rail === rows - 1 || rail === 0) change = -change
  }

  return r
}

function encrypt(string,rows) {
  let fence = [];
  for (let i = 0; i < rows; i++) fence.push([])
  let rail = 0;
  let change = 1;

  for (let char of string.split("")) {
    fence[rail].push(char)
    rail += change

    if (rail === rows - 1 || rail === 0) change = -change
  }

  let r = '';
  for (let rail of fence) r += rail.join("")

  return r
}

exports.encrypt = encrypt;
exports.decrypt = decrypt;

scripts/Caesar.js
function caesar(s) {
  var result = new Array();
  for (var j = 0; j < 26; j++) {
    var z = rot(s, j)
    result[j] = z;
  }
  return result
}

function rot(s, n) {
  var z = '';
  var w;
  for (var i = 0; i < s.length; i++) {
    if (s.charCodeAt(i) <= 90 && s.charCodeAt(i) >= 65) {
      w = (s.charCodeAt(i) + n - 65) % 26;
      w = String.fromCharCode(w + 65);
    } else if (s.charCodeAt(i) <= 122 && s.charCodeAt(i) >= 97) {
      w = (s.charCodeAt(i) + n - 97) % 26;
      w = String.fromCharCode(w + 97);
    } else {
      w = s[i];
    }
    z = z + w;
  }
  return z
}



function shift(s, max, d) {
  var result = new Array();
  for (var n = 0; n <= max; n++) {
    var z = ''
    for (var i = 0; i < s.length; i++) {
      if (d == 'r') {
        z += String.fromCharCode(s.charCodeAt(i) + n)
      } else {
        z += String.fromCharCode(s.charCodeAt(i) - n)
      }
    }
    result[n] = z;
  }
  return result
}

exports.rot = rot
exports.shift = shift
exports.caesar = caesar

scripts/Vigenere.js
// function $(id) {
//     return document.getElementById(id);
// }

function charOffset(char, offset) {
    if (offset < 0)
        offset += 26;
    if (/[a-z]/.test(char)) {
        return String.fromCharCode((char.charCodeAt(0) - 97 + offset) % 26 + 97);
    } else {
        return String.fromCharCode((char.charCodeAt(0) - 65 + offset) % 26 + 65);
    }
}

function Vigenere(strIn, key, encode) {
    var strOut = "";
    var j = 0; //  j 对应密钥key；
    for (var i = 0; i < strIn.length; i++) {
        var c = strIn[i];
        if (/[a-zA-Z]/.test(c)) {
            var offset = key.charCodeAt(j % key.length) - 97;
            j++;
            if (encode == false) offset = -offset;
            strOut += charOffset(c, offset);
        } else {
            strOut += c;
        }
    }
    return strOut;
}

// function enVigenere(argument) {
//     // body...
//     console.log("加密")
//     var plaintext = $("plaintext").value;
//     var key = $("key").value.toLowerCase(); //key仅为字母  这里没做判断
//     if (!key) { alert("请输入密钥"); return; }
//     $("ciphertext").value = Vigenere(plaintext, key, true);
// }

// function deVigenere(argument) {
//     // body...
//     console.log("解密")
//     var ciphertext = $("ciphertext").value;
//     var key = $("key").value.toLowerCase(); //key仅为字母  这里没做判断
//     if (!key) { alert("请输入密钥"); return; }
//     $("plaintext").value = Vigenere(ciphertext, key, false);
// }

function deVigenereAuto(ciphertext,best_len,minLen,maxLen) {
    // console.log("autokey");
    // var ciphertext = $("ciphertext").value;
    // var best_len = parseInt($("keyLen").value);
    var best_key = "";
    var count = [];
    var cipherMin = ciphertext.toLowerCase().replace(/[^a-z]/g, "");
    var freq = [8.167, 1.492, 2.782, 4.253, 12.702, 2.228, 2.015, 6.094, 6.966, 0.153, 0.772, 4.025, 2.406, 6.749, 7.507, 1.929, 0.095, 5.987, 6.327, 9.056, 2.758, 0.978, 2.360, 0.150, 1.974, 0.074];
    if (!best_len) {
        for (var best_len = minLen; best_len <= maxLen; best_len++) { //猜测key长度3————12
            var sum = 0;
            for (var j = 0; j < best_len; j++) {
                for (var i = 0; i < 26; i++) {
                    count[i] = 0;
                }
                for (var i = j; i < cipherMin.length; i += best_len) {
                    count[cipherMin[i].charCodeAt(0) - 97] += 1;
                }
                var ic = 0;
                var num = cipherMin.length / best_len;
                for (var i = 0; i < count.length; i++) {
                    ic += Math.pow(count[i] / num, 2);
                }
                sum += ic;
                // console.log(keyLen,ic);
            }
            console.log(best_len, sum / best_len);
            if (sum / best_len > 0.065) break; //确定密钥长度
        }
    }
    console.log(best_len)
    for (var j = 0; j < best_len; j++) {
        for (var i = 0; i < 26; i++) {
            count[i] = 0;
        }
        for (var i = j; i < cipherMin.length; i += best_len) {
            count[cipherMin[i].charCodeAt(0) - 97] += 1;
        }
        var max_dp = -1000000;
        var best_i = 0;

        for (var i = 0; i < 26; i++) {
            var cur_dp = 0.0;
            for (var k = 0; k < 26; k++) {
                cur_dp += freq[k] * count[(k + i) % 26]; //这里要找出频率分布匹配的key
            }
            if (cur_dp > max_dp) {
                max_dp = cur_dp;
                best_i = i;
            }
        }
        best_key += String.fromCharCode(best_i + 97);
    }
    // $("best_key").innerHTML = "最可能的密钥：" + best_key;
    // $("plaintext").value = Vigenere(ciphertext, best_key, false);
    console.log(best_key);
    var plaintext = Vigenere(ciphertext, best_key, false);
    return [plaintext,best_key,best_len]
}

// window.onload = function(argument) {
//     $("encryption").onclick = enVigenere;
//     $("decryption1").onclick = deVigenere;
//     $("decryption2").onclick = deVigenereAuto;
// }

exports.Vigenere = Vigenere
exports.deVigenereAuto = deVigenereAuto

scripts/brainfuck.js
var parse = (function () {
  var input;
  var output;
  var data;
  var ptr;
  var debug = false;

  var ops = {
    '+': function () {
      data[ptr] = data[ptr] || 0;
      data[ptr]++;
      debug && console.log('+', data[ptr], ptr);
    },

    '-': function () {
      data[ptr] = data[ptr] || 0;
      data[ptr]--;
      debug && console.log('-', data[ptr], ptr);
    },

    '<': function () {
      ptr--;
      if (ptr < 0) {
        ptr = 0; //Don't allow pointer to leave data array
      }
      debug && console.log('<', ptr);
    },

    '>': function () {
      ptr++;
      debug && console.log('>', ptr);
    },

    '.': function () {
      var c = String.fromCharCode(data[ptr]);
      output.push(c);
      debug && console.log('.', c, data[ptr]);
    },

    ',': function () {
      var c = input.shift();
      if (typeof c == "string") {
        data[ptr] = c.charCodeAt(0);
      }
      debug && console.log(',', c, data[ptr]);
    },
  };

  function program(nodes) {
    return function (inputString) {
      output = [];
      data = [];
      ptr = 0;

      input = inputString && inputString.split('') || [];

      nodes.forEach(function (node) {
        node();
      });

      return output.join('');
    }
  }

  function loop(nodes) {
    return function () {
      var loopCounter = 0;

      while(data[ptr] > 0) {
        if (loopCounter++ > 10000) { throw "Infinite loop detected"; }

        nodes.forEach(function (node) {
          node();
        });
      }
    };
  }



  var programChars;

  function parseProgram() {
    var nodes = [];
    var nextChar;

    while (programChars.length > 0) {
      nextChar = programChars.shift();
      if (ops[nextChar]) {
        nodes.push(ops[nextChar]);
      } else if (nextChar == '[') {
        nodes.push(parseLoop());
      } else if (nextChar == ']') {
        throw "Missing opening bracket";
      } else {
        // ignore it
      }
    }

    return program(nodes);
  }

  function parseLoop() {
    var nodes = [];
    var nextChar;

    while (programChars[0] != ']') {
      nextChar = programChars.shift();
      if (nextChar == undefined) {
        throw "Missing closing bracket";
      } else if (ops[nextChar]) {
        nodes.push(ops[nextChar]);
      } else if (nextChar == '[') {
        nodes.push(parseLoop());
      } else {
        // ignore it
      }
    }
    programChars.shift(); //discard ']'

    return loop(nodes);
  }

  function parse(str) {
    programChars = str.split('');
    return parseProgram();
  }

  return parse;
})();


function run(code, input) {
  return parse(code)(input);
}

exports.run = run


scripts/textManipulation.js
exports.upperString = function(text) {
	return text.toUpperCase();
}
exports.lowerString = function(text) {
	return text.toLowerCase();
}
exports.stripString = function(text) {
	return text.trim();
}
exports.reverseString = function(text) {
	return text.split("").reverse().join("");
}
exports.space2None = function(text) {
	return text.replace(/\s/g, "");
}
exports.space2Line = function(text) {
	return text.replace(/\s/g, "\n");
}
exports.convPath = function (text) {
	if (text.includes("\\")) {
		return text.replace(/\\/g, "/");	
	} else {
		return text.replace(/\//g, "\\");	
	}
}
exports.titleCase = function(text) {
	return text.toLowerCase().replace(/( |^)[a-z]/g, (L) => L.toUpperCase()); 
}
exports.addQuotByComma = function(text) {
	return text.split(',').map(i => { return '"' + i.trim() + '"' }).join(", ");
}
exports.addQuotBySpace = function(text) {
	return text.split(' ').map(i => { return '"' + i.trim() + '"' }).join(" ");
}
exports.stringLen = function(text) {
	return text.length;
}

extension.js
const vscode = require('vscode')
const caesar = require('./scripts/Caesar')
const vigenere = require('./scripts/Vigenere')
const morse = require('xmorse')
const basic = require('./scripts/Basic')
const manip = require('./scripts/textManipulation')
const crypto = require('crypto')
const fs = require('fs')
const fence = require('./scripts/Fence')
const jssm4 = require('jssm4')

/**
 * 插件被激活时触发，所有代码总入口
 * @param {*} context 插件上下文
 */

//格式化
String.prototype.format = function () {
  var args = arguments
  return this.replace(/\{(\d+)\}/gm, (ms, p1) => {
    return typeof args[p1] == 'undefined' ? ms : args[p1]
  })
}

//显示信息与更新文本
function outputChannelShow(text) {
  outputChannel.show()
  outputChannel.appendLine(text)
}

function outputChannelUpdate(alg, before, after, update) {
  if (after) {
    //output
    outputChannelShow('------------------------------------')
    outputChannelShow('  ◆   ' + alg + ':')
    outputChannelShow('[ ⇐ ] ' + before)
    outputChannelShow('[ ⇒ ] ' + after)
    //update
    if (update) {
      editor.edit((editBuilder) => {
        editBuilder.replace(selection, after)
      })
      outputChannelShow('[ ✔ ] 文本已自动更新，按 Ctrl + Z 撤销')
    }
    outputChannelShow('------------------------------------')
  } else {
    vscode.window.showErrorMessage('转换失败！请检查文本类型或查看日志！')
  }
}

//凯撒
async function caesarCipher(text) {
  result = caesar.caesar(text)
  console.log(result)
  let choise = await vscode.window.showQuickPick(result, {
    placeHolder: '选择其中一种结果进行替换',
  })
  if (choise) {
    vscode.window.showInformationMessage(
      '选择了位移了' + result.indexOf(choise) + '位的结果',
    )
    outputChannelUpdate('caesarCipher', text, choise, true)
  }
}
//位移
async function characterOffset(text) {
  let value = await vscode.window.showInputBox({
    prompt: '输入最大的位移位数n和方向l或r，以逗号隔开',
    value: '50,l',
  })
  if (value) {
    let n = value.split(',')[0]
    let d = value.split(',')[1]
    result = caesar.shift(text, n, d)
    console.log(result)
    let choise = await vscode.window.showQuickPick(result, {
      placeHolder: '选择其中一种结果进行替换',
    })
    if (choise) {
      vscode.window.showInformationMessage(
        '选择了位移了' + result.indexOf(choise) + '位的结果',
      )
      outputChannelUpdate('characterOffset', text, choise, true)
    }
  }
}
//栅栏
async function fenceCipher(text, alg) {
  let result = []
  if (alg.includes('Encode')) {
    for (var i = 2; i < text.length; i++) {
      result.push(fence.encrypt(text, i))
    }
  } else {
    for (var i = 2; i < text.length; i++) {
      result.push(fence.decrypt(text, i))
    }
  }
  console.log(result)
  let choise = await vscode.window.showQuickPick(result, {
    placeHolder: '选择其中一种结果进行替换',
  })
  if (choise) {
    vscode.window.showInformationMessage(
      '选择了栏数为' + (result.indexOf(choise) + 2) + '的结果',
    )
    outputChannelUpdate('fenceCipher', text, choise, true)
  }
}

//维吉尼亚
async function vigenereCipher(text, alg) {
  let key = await vscode.window.showInputBox({
    placeHolder: '输入秘钥',
  })
  if (key) {
    if (alg.includes('Encode')) {
      var result = vigenere.Vigenere(text, key, true)
    } else {
      var result = vigenere.Vigenere(text, key, false)
    }
    outputChannelUpdate(alg, text, result, true)
  }
}

async function vigenereAutoDecode(text) {
  let minlen = 1 //从1开始猜测key的长度
  while (true) {
    let guessKey = new Promise((resolve) => {
      let result = vigenere.deVigenereAuto(text, false, minlen, 100) //默认key的长度未知
      resolve(result)
    })
    result = await guessKey
    outputChannelUpdate('vigenereAutoDecode', text, result[0], false)
    outputChannelShow('最有可能的key为：' + result[1])
    minlen = result[2] + 1 //如果没猜对，从猜出的秘钥长度+1之后继续猜
    let choise = await vscode.window.showInformationMessage(
      '请在输出日志中确认秘钥和明文是否猜测正确？',
      '是',
      '否',
    )
    if (choise != '否') {
      outputChannelUpdate('vigenereAutoDecode', text, result[0], true)
      break
    }
  }
}

function cryptoCipher(plainText, algorithm, key, iv, encoding = 'base64') {
  let cipher = crypto.createCipheriv(algorithm, key, iv)
  return cipher.update(plainText, 'utf8', encoding) + cipher.final(encoding)
}

function cryptoDecipher(cipher, algorithm, key, iv, encoding = 'base64') {
  let decipher = crypto.createDecipheriv(algorithm, key, iv)
  return decipher.update(cipher, encoding, 'utf8') + decipher.final('utf8')
}
let latestSymmetricCryptionKeyIv
//对称加密
async function symmetricCryption(text, alg) {
  let ciphers = crypto.getCiphers()
  ciphers.includes('sm4-ecb') || ciphers.push('sm4-ecb')
  let algorithm = await vscode.window.showQuickPick(ciphers, {
    placeHolder: '请选择要使用的对称加密算法',
  })
  if (!algorithm) return
  let value = await vscode.window.showInputBox({
    placeHolder: '请输入秘钥和初始向量(如有),并以英文“,”隔开',
    value: latestSymmetricCryptionKeyIv || '',
  })
  latestSymmetricCryptionKeyIv = value
  let key = value.split(',')[0] || ''
  let iv = value.split(',')[1] || ''
  outputChannelShow('Key: ' + key + ' IV: ' + iv)
  let result
  try {
    if (alg.includes('Encryption')) {
      result =
        algorithm === 'sm4-ecb'
          ? new jssm4(key).encryptData_ECB(text)
          : cryptoCipher(text, algorithm, key, iv)
    } else {
      result =
        algorithm === 'sm4-ecb'
          ? new jssm4(key).decryptData_ECB(text)
          : cryptoDecipher(text, algorithm, key, iv)
    }
  } catch (err) {
    outputChannelShow('[ ✘ ] ' + err)
  }
  outputChannelUpdate(alg + ' - ' + algorithm, text, result, true)
}

//RSA加密
function checkKey(keyfile, key) {
  let pri = [crypto.privateEncrypt, crypto.privateDecrypt]
  let pub = [crypto.publicEncrypt, crypto.publicDecrypt]
  if (key.includes('PRIVATE')) {
    outputChannelShow('选择了私钥文件: ' + keyfile)
    return pri
  } else if (key.includes('PUBLIC')) {
    outputChannelShow('选择了公钥文件: ' + keyfile)
    return pub
  } else {
    outputChannelShow('[ ✘ ] 秘钥格式似乎不对')
    outputChannelShow('私钥的开头和结尾分别为:')
    outputChannelShow('-----BEGIN RSA PRIVATE KEY-----')
    outputChannelShow('-----END RSA PRIVATE KEY-----')
    outputChannelShow('公钥的开头和结尾分别为:')
    outputChannelShow('-----BEGIN PUBLIC KEY-----')
    outputChannelShow('-----END PUBLIC KEY-----')
  }
}

async function rsaCryption(text, alg) {
  try {
    let file = await vscode.window.showOpenDialog({
      openLabel: '选择秘钥',
    })
    if (file) {
      let keyfile = file[0].fsPath
      let key = fs.readFileSync(keyfile).toString('utf-8')
      if (alg.includes('Encryption')) {
        crypt = checkKey(keyfile, key)[0]
        var result = crypt(key, new Buffer(text)).toString('base64')
      } else {
        crypt = checkKey(keyfile, key)[1]
        var result = crypt(key, new Buffer(text, 'base64')).toString('utf8')
      }
      outputChannelUpdate(alg, text, result, true)
    }
  } catch (err) {
    outputChannelShow(err)
  }
}

//摩斯
async function morseCoding(text, alg) {
  let value = await vscode.window.showInputBox({
    value: "{ space: '/', long: '-', short: '.' }",
    prompt: '请输入对应的间隔符、划、点的符号',
  })
  if (value) {
    let option = eval('(' + value + ')')
    if (alg.includes('Encode')) {
      var result = morse.encode(text, option)
    } else {
      var result = morse.decode(text, option)
    }
    outputChannelUpdate(alg, text, result, true)
  }
}

//十六进制计算器
async function calculator(text) {
  let value = await vscode.window.showInputBox({
    placeHolder: '输入要计算的公式，非十进制数以0x、0b、0o开头',
    value: text,
  })
  if (value) {
    try {
      let result = eval(value)
      let dec = parseFloat(result)
      let hex = '0x' + dec.toString(16)
      let oct = '0o' + dec.toString(8)
      let bin = '0b' + dec.toString(2)
      let str =
        /^[\x20-\x7e]+$/.test(basic.number2string(dec)) &&
        basic.number2string(dec)
      log = `dec: {0}
			hex: {1}
			bin: {2}
			oct: {3}`.format(dec, hex, bin, oct)
      if (str) {
        log += '\n			str: "{0}"'.format(str)
      }
      outputChannelUpdate('hexadecimalConverter', value, log, false)
    } catch (err) {
      vscode.window.showErrorMessage('输入非公式，将先转换成十六进制再计算')
      let hex = '0x' + basic.string2hex(value)
      calculator(hex)
    }
  }
}

//处理text/x-code-output兼容问题
//获取output channel的高亮语法
function getPatterns(file) {
  let content = fs.readFileSync(file, 'utf-8')
  let patterns = /<key>patterns<\/key>[\s|\S]*?<array>([\s|\S]*?)<\/array>/.exec(
    content,
  )[1]
  return patterns
}

//去除相关插件JSON文件里的"text/x-code-output"，以免冲突
function rmOutput(file) {
  let content = fs.readFileSync(file, 'utf-8')
  let New = content.replace('text/x-code-output', 'text/bak-x-code-output')
  fs.writeFileSync(file, New)
}

//获取定义了output channel语法的文件，并整合语法
function getLang() {
  let lang = ''
  const extension = vscode.extensions.all
  for (let e of extension) {
    try {
      let mimetypes = e.packageJSON.contributes.languages[0].mimetypes
      if (mimetypes.includes('text/x-code-output')) {
        let id = e.id
        if (id != 'fofolee.crypto-tools') {
          let extensionPath = e.extensionPath
          let grammarsPath = e.packageJSON.contributes.grammars[0].path.substr(
            1,
          )
          rmOutput(manip.convPath(extensionPath + '/package.json'))
          outputChannelShow('发现定义了output语法的冲突插件:\n' + extensionPath)
          file = manip.convPath(extensionPath + grammarsPath)
          console.log(file)
          lang +=
            '        <!-- ' +
            id +
            ' start -->' +
            getPatterns(file) +
            '<!-- ' +
            id +
            ' end -->\n'
        }
      }
    } catch (err) {}
  }
  return lang
}

outputChannel = vscode.window.createOutputChannel('crypto')

exports.activate = (context) => {
  // 注册命令
  context.subscriptions.push(
    vscode.commands.registerCommand('crypto.EncodeDecode', async () => {
      editor = vscode.window.activeTextEditor
      selection = editor.selection
      let text = editor.document.getText(selection).trim()
      let algorithm = await vscode.window.showQuickPick(
        [
          {
            label: 'Base64/32/16 Decode',
            detail: '自动进行(url-safe)base64/32/16解密',
            target: basic.baseDecode,
          },
          {
            label: 'Base64 Encode',
            detail: 'base64加密',
            target: basic.base64Encode,
          },
          {
            label: 'Url-Safe Base64 Encode',
            detail: 'url-safe base64加密',
            target: basic.urlSafeBase64Encode,
          },
          {
            label: 'Base32 Encode',
            detail: 'base32加密',
            target: basic.base32Encode,
          },
          {
            label: 'Base16 Encode',
            detail: 'base16加密',
            target: basic.base16Encode,
          },
          {
            label: 'MD5',
            detail: 'MD5哈希算法',
            target: basic.md5Hash,
          },
          {
            label: 'SHA512',
            detail: 'SHA512哈希算法',
            target: basic.sha512Hash,
          },
          {
            label: 'Url Decode',
            detail: 'url加密',
            target: basic.urlDecode,
          },
          {
            label: 'Url Encode',
            detail: 'url解密',
            target: basic.urlEncode,
          },
          {
            label: 'Html Entities',
            detail: 'html加密',
            target: basic.htmlEncode,
          },
          {
            label: 'Html Entity Decode',
            detail: 'html解密',
            target: basic.htmlDecode,
          },
          {
            label: 'Morse Encode',
            detail: '摩斯密码加密',
            target: morseCoding,
          },
          {
            label: 'Morse Decode',
            detail: '摩斯密码解密',
            target: morseCoding,
          },
          {
            label: 'ROT13',
            detail: 'ROT13加密',
            target: basic.rot13,
          },
          {
            label: 'Quote-Printable Decode',
            detail: 'Quote-Printable解密',
            target: basic.quopriDecode,
          },
          {
            label: 'Quote-Printable Encode',
            detail: 'Quote-Printable加密',
            target: basic.quopriEncode,
          },
          {
            label: 'Bubble Babble Decode',
            detail: 'Bubble Babble 解密',
            target: basic.bubbleDecode,
          },
          {
            label: 'Bubble Babble Encode',
            detail: 'Bubble Babble 加密',
            target: basic.bubbleEncode,
          },
          {
            label: 'Brainfuck Decode',
            detail: 'Brainfuck 解密',
            target: basic.brainfuckDecode,
          },
          {
            label: 'Number To String',
            detail: '整型转字符',
            target: basic.number2string,
          },
          {
            label: 'String To Number',
            detail: '字符转整型',
            target: basic.string2number,
          },
          {
            label: 'String To Hex',
            detail: '字符转十六进制',
            target: basic.string2hex,
          },
          {
            label: 'Hex To String',
            detail: '十六进制转字符',
            target: basic.hex2string,
          },
          {
            label: 'String To Bin',
            detail: '字符转二进制',
            target: basic.string2bin,
          },
          {
            label: 'Bin To String',
            detail: '二进制转字符',
            target: basic.bin2string,
          },
        ],
        {
          placeHolder: '选择一种算法',
        },
      )
      if (algorithm) {
        let result = algorithm.target(text)
        outputChannelUpdate(algorithm.label, text, result, true)
      }
    }),
  )

  context.subscriptions.push(
    vscode.commands.registerCommand('crypto.EncryptDecrypt', async () => {
      editor = vscode.window.activeTextEditor
      selection = editor.selection
      let text = editor.document.getText(selection).trim()
      let algorithm = await vscode.window.showQuickPick(
        [
          {
            label: 'Symmetric Decryption',
            detail: '对称密码解密算法，密文格式为Base64',
            target: symmetricCryption,
          },
          {
            label: 'Symmetric Encryption',
            detail: '对称密码加密算法，密文格式为Base64',
            target: symmetricCryption,
          },
          {
            label: 'RSA Decryption',
            detail: '使用公钥或私钥文件进行RSA解密',
            target: rsaCryption,
          },
          {
            label: 'RSA Encryption',
            detail: '使用公钥或私钥文件进行RSA加密',
            target: rsaCryption,
          },
          {
            label: 'Caesar Cipher',
            detail: '凯撒密码',
            target: caesarCipher,
          },
          {
            label: 'Character Offset',
            detail: '字符位移',
            target: characterOffset,
          },
          {
            label: 'Fence Decode',
            detail: '栅栏密码解密',
            target: fenceCipher,
          },
          {
            label: 'Fence Encode',
            detail: '栅栏密码加密',
            target: fenceCipher,
          },
          {
            label: 'Vigenere Encode',
            detail: '维基尼亚加密',
            target: vigenereCipher,
          },
          {
            label: 'Vigenere Decode',
            detail: '维基尼亚解密',
            target: vigenereCipher,
          },
          {
            label: 'Vigenere Decode with No Key',
            detail: '维基尼亚无秘钥解密',
            target: vigenereAutoDecode,
          },
        ],
        {
          placeHolder: '选择一种算法',
        },
      )
      if (algorithm) {
        algorithm.target(text, algorithm.label)
      }
    }),
  )

  context.subscriptions.push(
    vscode.commands.registerCommand('crypto.textManipulation', async () => {
      editor = vscode.window.activeTextEditor
      selection = editor.selection
      let text = editor.document.getText(selection).trim()
      let algorithm = await vscode.window.showQuickPick(
        [
          {
            label: 'Reverse String',
            detail: '字符串逆转',
            target: manip.reverseString,
          },
          {
            label: 'Upper Case',
            detail: '全部大写',
            target: manip.upperString,
          },
          {
            label: 'Lower Case',
            detail: '全部小写',
            target: manip.lowerString,
          },
          {
            label: 'Strip String',
            detail: '去除左右空格、换行',
            target: manip.stripString,
          },
          {
            label: 'Space To None',
            detail: '去除所有空格',
            target: manip.space2None,
          },
          {
            label: 'Space To Line',
            detail: '空格转换行',
            target: manip.space2Line,
          },
          {
            label: 'Convert Path',
            detail: '\\和/互转',
            target: manip.convPath,
          },
          {
            label: 'Title Case',
            detail: '所有词首字母大写',
            target: manip.titleCase,
          },
          {
            label: 'String Lenght',
            detail: '获取文本长度',
            target: manip.stringLen,
          },
          {
            label: 'Add Quot By Comma',
            detail: '为每个以逗号分隔的词加上双引号',
            target: manip.addQuotByComma,
          },
          {
            label: 'Add Quot By Space',
            detail: '为每个以空格分隔的词加上双引号',
            target: manip.addQuotBySpace,
          },
        ],
        {
          placeHolder: '选择一种处理方案',
        },
      )
      if (algorithm) {
        let result = algorithm.target(text)
        if (algorithm.label != 'String Lenght') {
          outputChannelUpdate(algorithm.label, text, result, true)
        } else {
          outputChannelUpdate(algorithm.label, text, result, false)
          vscode.window.showInformationMessage('长度为：' + result)
        }
      }
    }),
  )

  context.subscriptions.push(
    vscode.commands.registerCommand('crypto.hexadecimalCalculator', () => {
      calculator('')
    }),
  )

  context.subscriptions.push(
    vscode.commands.registerCommand('crypto.outputColorPatch', async () => {
      let choise = await vscode.window.showInformationMessage(
        '如果安装本插件后，本插件的输出没有高亮，或者造成其他插件的输出没有高亮，则是因为和其他一些插件的输出语法产生了冲突，是否尝试自动修复这些冲突？',
        '是',
        '否',
      )
      if (choise == '是') {
        outputChannelShow('正在尝试寻找问题···')
        let langFile = __dirname + '/syntaxes/crypto-tools-output.tmLanguage'
        let file = fs.readFileSync(langFile, 'utf-8')
        let lang = getLang()
        let New = file.replace('        </array>', lang + '        </array>')
        console.log(New)
        fs.writeFileSync(langFile, New)
        if (lang) {
          outputChannelShow('语法文件已整合，请重启编辑器！')
        } else {
          outputChannelShow('未发现冲突的文件！')
        }
      }
    }),
  )
}
